File: src/main/java/model/FamilyModel.java
package model;


import com.koenig.commonModel.Component;
import com.koenig.commonModel.Family;
import com.koenig.commonModel.User;
import com.koenig.communication.ConnectUtils;
import com.koenig.communication.messages.FamilyMessage;
import com.koenig.communication.messages.TextMessage;
import com.koenig.communication.messages.family.CreateUserMessage;
import com.koenig.communication.messages.family.FamilyMemberMessage;
import com.koenig.communication.messages.family.FamilyTextMessages;
import com.koenig.communication.messages.family.UserMessage;
import communication.OnReceiveMessageListener;
import communication.Server;
import database.UserDatabase;
import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.SQLException;
import java.util.List;

public class FamilyModel implements OnReceiveMessageListener {

    private Logger logger = LoggerFactory.getLogger(getClass().getSimpleName());
    private Server server;
    private UserDatabase database;
    private FinanceModel financeModel;

    public void start(UserDatabase userDatabase) throws SQLException {
        logger.info("Start");
        database = userDatabase;

        database.start();
        server = new Server(ConnectUtils.PORT, this);
        financeModel = new FinanceModel(server, new FamilyConnectionService(database));
        server.start();

    }


    @Override
    public void onReceiveMessage(FamilyMessage message) {
        try
        {
            logger.info("Received message: " + message.getName());
            switch (message.getComponent()) {

                case FINANCE:
                    financeModel.onReceiveMessage(message);
                    break;
                case CONTRACTS:
                    break;
                case OWNINGS:
                    break;
                case HEALTH:
                    break;
                case WIKI:
                    break;
                case FAMILY:
                    onReceiveFamilyMessage(message);
                    break;
                case WORK:
                    break;
            }


        } catch (Exception e) {
            logger.error("Error while working with message: " + e.getMessage());
        }
    }

    private void onReceiveFamilyMessage(FamilyMessage message) {
        String userId = message.getFromId();
        switch (message.getName()) {
            case TextMessage.NAME:
                processCommands(((TextMessage) message).getText(), message.getFromId());
                break;

            case CreateUserMessage.NAME:
                CreateUserMessage createUserMessage = (CreateUserMessage) message;
                addUserToDatabase(createUserMessage.getUserName().trim(), createUserMessage.getBirthday(), userId);
                break;
        }

    }

    private void processCommands(String text, String fromId) {
        String[] words = text.split(FamilyMessage.SEPARATOR);
        switch (words[0]) {
            case FamilyTextMessages.CREATE_FAMILY:
                String familyName = words[1].trim();
                logger.info("Creating new family: " + familyName);
                if (createNewFamily(familyName, fromId)) {
                    logger.info("Created new family: " + familyName);
                    joinFamily(familyName, fromId);
                }

                break;
            case FamilyTextMessages.JOIN_FAMILY:
                familyName = words[1].trim();
                joinFamily(familyName, fromId);
                break;

            case FamilyTextMessages.LOGIN:
                login(fromId);
                break;

            case FamilyTextMessages.GET_FAMILY_MEMBER:
                sendFamilyMembers(fromId);
                break;
            default:
                logger.error("Unknown Command: " + text);
                break;
        }
    }

    private void sendFamilyMembers(String userId) {
        try {
            List<User> users = database.getFamilyMemberFrom(userId);
            server.sendMessage(new FamilyMemberMessage(users), userId);
        } catch (SQLException e) {
            logger.error(e.getMessage());
            sendFamilyCommand(FamilyTextMessages.GET_FAMILY_MEMBER_FAIL, userId);
        }

    }

    private void login(String userId) {
        try {
            User user = database.getUserById(userId);
            if (user == null) {
                throw new SQLException("User does not exist with id: " + userId);
            }

            server.sendMessage(new UserMessage(user), userId);
            logger.info(user.getName() + " logged in(Family: " + user.getFamily() + ")");
        } catch (SQLException e) {
            logger.error("Couldn't find user: " + e.getMessage());
            sendFamilyCommand(FamilyTextMessages.LOGIN_FAIL, userId);
        }
    }

    private boolean createNewFamily(String familyName, String userId) {
        try {
            // check if family already exists, name must be unique
            Family family = database.getFamilyByName(familyName);
            if (family != null) {
                throw new SQLException("Family exists already");
            }

            database.addFamily(new Family(familyName), userId);
            // TODO: create file database for each new family

            sendFamilyCommand(FamilyTextMessages.CREATE_FAMILY_SUCCESS, userId);
            return true;

        } catch (SQLException e) {
            logger.error("Couldn't add family" + e.getMessage());
            server.sendMessage(new TextMessage(Component.FAMILY, FamilyTextMessages.CREATE_FAMILY_FAIL), userId);
            return false;
        }
    }

    private void sendFamilyCommand(String command, String userId) {
        server.sendMessage(new TextMessage(Component.FAMILY, command), userId);
    }

    private void joinFamily(String familyName, String userId) {
        // TODO: if first then its the admin
        try {
            database.addUserToFamily(familyName, userId);
            sendFamilyCommand(FamilyTextMessages.JOIN_FAMILY_SUCCESS, userId);
            logger.info("Famile " + familyName + " beigetreten");
        } catch (SQLException e) {
            logger.error("Couldn't join user to family: " + e.getMessage());
            server.sendMessage(new TextMessage(Component.FAMILY, FamilyTextMessages.JOIN_FAMILY_FAIL), userId);
        }
    }

    private void addUserToDatabase(String userName, DateTime birthday, String userId) {
        try {
            User user = new User(userId, userName.trim(), "", birthday);
            database.addUser(user, userId);
            logger.info("Adding user " + userName + ", Birthday: " + birthday);
            server.sendMessage(new TextMessage(Component.FAMILY, FamilyTextMessages.CREATE_USER_SUCCESS), userId);
        } catch (SQLException e) {
            logger.error("Couldn't add user to database: " + e.getMessage());
            server.sendMessage(new TextMessage(Component.FAMILY, FamilyTextMessages.CREATE_USER_FAIL), userId);
        }
    }


    public void stop() {
        server.stop();
        try {
            database.stop();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }


}

File: src/main/java/model/FamilyModel.kt
package model


import com.koenig.commonModel.Component
import com.koenig.commonModel.Family
import com.koenig.commonModel.User
import com.koenig.communication.ConnectUtils
import com.koenig.communication.messages.FamilyMessage
import com.koenig.communication.messages.TextMessage
import com.koenig.communication.messages.family.CreateUserMessage
import com.koenig.communication.messages.family.FamilyMemberMessage
import com.koenig.communication.messages.family.FamilyTextMessages
import com.koenig.communication.messages.family.UserMessage
import communication.OnReceiveMessageListener
import communication.Server
import database.FamilyDbRepository
import database.UserDatabase
import model.finance.FinanceModel
import org.joda.time.DateTime
import org.joda.time.YearMonth
import org.slf4j.LoggerFactory
import java.sql.SQLException
import java.util.*

class FamilyModel : OnReceiveMessageListener {

    private val logger = LoggerFactory.getLogger(javaClass.simpleName)
    private var server: Server = Server(ConnectUtils.PORT, this)
    private lateinit var userDatabase: UserDatabase
    lateinit var familyConnectionService: FamilyConnectionService
    private lateinit var financeModel: FinanceModel

    @Throws(SQLException::class)
    fun start(userDatabase: UserDatabase) {
        logger.info("Start")
        this.userDatabase = userDatabase
        familyConnectionService = FamilyConnectionService(this.userDatabase)
        financeModel = FinanceModel(server, familyConnectionService, userDatabase.userService, FamilyDbRepository(this.userDatabase.familyTable))
        userDatabase.start()
        server.start()
        val thomasId = "c572d4e7-da4b-41d8-9c1f-7e9a97657155"
        val thomas = this.userDatabase.getUserById(thomasId)
        val userthomas = User(thomasId, "Thomas", "König", DateTime(1987, 6, 14, 12, 0))
        if (thomas == null) {
            this.userDatabase.addUser(userthomas, thomasId)
        }
        val milenaId = "c6540de0-46bb-42cd-939b-ce52677fa19d"
        val milena = this.userDatabase.getUserById(milenaId)
        val userMilena = User(milenaId, "Milena", "König", DateTime(1987, 8, 10, 12, 0))
        if (milena == null) {
            this.userDatabase.addUser(userMilena, milenaId)
        }

        val family = this.userDatabase.getFamilyByName("König")
        if (family == null) {
            val users = ArrayList<User>(2)
            users.add(userMilena)
            users.add(userthomas)
            this.userDatabase.addFamily(Family("König", users, YearMonth(2015, 1)), thomasId)
        }

    }


    @Synchronized
    override fun onReceiveMessage(message: FamilyMessage) {
        try {
            logger.info("Received message: " + message.name)
            when (message.component) {

                Component.FINANCE -> financeModel.onReceiveMessage(message)
                Component.CONTRACTS -> {
                }
                Component.OWNINGS -> {
                }
                Component.HEALTH -> {
                }
                Component.WIKI -> {
                }
                Component.FAMILY -> onReceiveFamilyMessage(message)
                Component.WORK -> {
                }
            }


        } catch (e: Exception) {
            logger.error("Error while working with message: " + e.message)
        }

    }

    private fun onReceiveFamilyMessage(message: FamilyMessage) {
        val userId = message.fromId
        when (message.name) {
            TextMessage.NAME -> processCommands((message as TextMessage).text, message.fromId)

            CreateUserMessage.NAME -> {
                val createUserMessage = message as CreateUserMessage
                addUserToDatabase(createUserMessage.userName.trim { it <= ' ' }, createUserMessage.birthday, userId)
            }
        }

    }

    private fun processCommands(text: String, fromId: String) {
        val words = text.split(FamilyMessage.SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var familyName = words[1].trim { it <= ' ' }
        when (words[0]) {
            FamilyTextMessages.CREATE_FAMILY -> {

                logger.info("Creating new family: $familyName")
                if (createNewFamily(familyName, fromId)) {
                    logger.info("Created new family: $familyName")
                    joinFamily(familyName, fromId)
                }
            }
            FamilyTextMessages.JOIN_FAMILY -> {
                familyName = words[1].trim { it <= ' ' }
                joinFamily(familyName, fromId)
            }

            FamilyTextMessages.LOGIN -> login(fromId)

            FamilyTextMessages.GET_FAMILY_MEMBER -> sendFamilyMembers(fromId)
            else -> logger.error("Unknown Command: $text")
        }
    }

    private fun sendFamilyMembers(userId: String) {
        try {
            val users = userDatabase.getFamilyMemberFrom(userId)
            server.sendMessage(FamilyMemberMessage(users), userId)
        } catch (e: SQLException) {
            logger.error(e.message)
            sendFamilyCommand(FamilyTextMessages.GET_FAMILY_MEMBER_FAIL, userId)
        }

    }

    private fun login(userId: String) {
        try {
            val user = userDatabase.getUserById(userId) ?: throw SQLException("User does not exist with id: $userId")

            server.sendMessage(UserMessage(user), userId)
            logger.info(user.name + " logged in(Family: " + user.family + ")")
        } catch (e: SQLException) {
            logger.error("Couldn't find user: " + e.message)
            sendFamilyCommand(FamilyTextMessages.LOGIN_FAIL, userId)
        }

    }

    private fun createNewFamily(familyName: String, userId: String): Boolean {
        try {
            // check if family already exists, name must be unique
            val family = userDatabase.getFamilyByName(familyName)
            if (family != null) {
                throw SQLException("Family exists already")
            }

            userDatabase.addFamily(Family(familyName), userId)
            // TODO: create file userDatabase for each new family

            sendFamilyCommand(FamilyTextMessages.CREATE_FAMILY_SUCCESS, userId)
            return true

        } catch (e: SQLException) {
            logger.error("Couldn't add family" + e.message)
            server.sendMessage(TextMessage(Component.FAMILY, FamilyTextMessages.CREATE_FAMILY_FAIL), userId)
            return false
        }

    }

    private fun sendFamilyCommand(command: String, userId: String) {
        server.sendMessage(TextMessage(Component.FAMILY, command), userId)
    }

    private fun joinFamily(familyName: String, userId: String) {
        // TODO: if first then its the admin
        try {
            userDatabase.addUserToFamily(familyName, userId)
            sendFamilyCommand(FamilyTextMessages.JOIN_FAMILY_SUCCESS, userId)
            logger.info("Famile $familyName beigetreten")
        } catch (e: SQLException) {
            logger.error("Couldn't join user to family: " + e.message)
            server.sendMessage(TextMessage(Component.FAMILY, FamilyTextMessages.JOIN_FAMILY_FAIL), userId)
        }

    }

    private fun addUserToDatabase(userName: String, birthday: DateTime, userId: String) {
        try {
            if (userName.trim { it <= ' ' }.isEmpty()) throw SQLException("Empty name not allowed")
            val user = User(userId, userName.trim { it <= ' ' }, "", birthday)
            userDatabase.addUser(user, userId)
            logger.info("Adding user $userName, Birthday: $birthday")
            server.sendMessage(TextMessage(Component.FAMILY, FamilyTextMessages.CREATE_USER_SUCCESS), userId)
        } catch (e: SQLException) {
            logger.error("Couldn't add user to userDatabase: " + e.message)
            server.sendMessage(TextMessage(Component.FAMILY, FamilyTextMessages.CREATE_USER_FAIL), userId)
        }

    }


    fun stop() {
        server.stop()
        try {
            userDatabase.stop()
        } catch (e: SQLException) {
            e.printStackTrace()
        }

    }


}
